src/response.rs

```diff
@@ -38,81 +39,170 @@ impl<'a> Response<'a> {
+
+        iterator
+    }
+
+    /// Get the response body
+    pub fn body<'conn, C: Read>(self, conn: &'conn mut C) -> ResponseBody<'conn, C>
+    where
+        'a: 'conn,
+    {
+        // Move the body part of the bytes in the header buffer to the beginning of the buffer
+        let header_buf = self.header_buf;
+        for i in 0..self.body_pos {
+            header_buf[i] = header_buf[self.header_len + i];
+        }
+
+        // The header buffer is now the body buffer
+        let body_buf = header_buf;
+        let reader = BodyReader {
+            conn,
+            remaining: self.content_length.map(|cl| cl - self.body_pos),
         };
-        //trace!("HTTP response: {:?}", response);
-        Ok(response)
+
+        ResponseBody {
+            body_buf,
+            body_pos: self.body_pos,
+            reader,
+        }
+    }
+}
+
+pub struct HeaderIterator<'a>(usize, [httparse::Header<'a>; 64]);
+
+impl<'a> Iterator for HeaderIterator<'a> {
+    type Item = (&'a str, &'a [u8]);
+
+    fn next(&mut self) -> Option<Self::Item> {
+        let result = self.1.get(self.0);
+
+        self.0 += 1;
+
+        result.map(|h| (h.name, h.value))
+    }
+}
+
+/// Response body
+///
+/// This type contains the original header buffer provided to `read_headers`,
+/// now renamed to `body_buf`, the number of read body bytes that are available
+/// in `body_buf`, and a reader to be used for reading the remaining body.
+pub struct ResponseBody<'a, C: Read> {
+    /// The buffer initially provided to read the header.
+    pub body_buf: &'a mut [u8],
+    /// The number bytes raed from the body and available in `body_buf`.
+    pub body_pos: usize,
+    /// The reader to be used for reading the remaining body.
+    pub reader: BodyReader<'a, C>,
+}
+
+impl<'a, C: Read> ResponseBody<'a, C> {
+    /// Read the entire body
+    pub async fn read_to_end(mut self) -> Result<&'a [u8], Error> {
+        // Read into the buffer after the portion that was already received when parsing the header
+        let len = self.reader.read_to_end(&mut self.body_buf[self.body_pos..]).await?;
+        Ok(&self.body_buf[..self.body_pos + len])
+    }
+}
+
+/// Response body reader
+pub struct BodyReader<'a, C: Read> {
+    conn: &'a mut C,
+    remaining: Option<usize>,
+}
+
+impl<C: Read> BodyReader<'_, C> {
+    /// Read until the end of the body
+    pub async fn read_to_end(&mut self, buf: &mut [u8]) -> Result<usize, Error> {
+        let to_read = self.remaining.ok_or(Error::Codec)?;
+        if buf.len() < to_read {
+            // The buffer is not sufficiently large to contain the entire body
+            return Err(Error::BufferTooSmall);
+        }
+
+        self.conn.read_exact(&mut buf[..to_read]).await.map_err(|e| match e {
+            ReadExactError::UnexpectedEof => Error::Network(ErrorKind::Other),
+            ReadExactError::Other(e) => e.kind().into(),
+        })?;
+
+        self.remaining.replace(0);
+
+        Ok(to_read)
+    }
```

Comment:
I think this should be split into a fn read() that reads up to min(buf.len(), self.remaining) and returns the number of bytes read, and another fn read_to_end() which 'loops' over that to fill the entire buffer.